src/chains/ethereum/ethereum/src/block-time.ts

```diff
@@ -0,0 +1,100 @@
+export interface BlockTime {
+  /**
+   * Set the current time of the BlockTime instance. Mining immediately afterwards will result in a block timestamp of ~timestamp (some milliseconds could pass before the block is mined).
+   * @param  {number | Date} timestamp - The time that should be considered "current time".
+   * @returns number - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setTime(timestamp: number | Date): number;
+
+  /**
+   * Get the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @returns number - The "offset" for the BlockTime instance in milliseconds.
+   */
+  getOffset(): number;
+
+  /**
+   * Set the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @param {number} offset - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setOffset(offset: number): void;
+
+  /**
+   * Create a timestamp for a new block in seconds. Can specify the timestamp for the block, in milliseconds, which will be reflected in future calls to this function.
+   * @param  {number} timestamp? - Optional - The timestamp for the block that will be mined, in milliseconds.
+   * @returns number - the timestamp in seconds
+   */
+  createBlockTimestampInSeconds(timestamp?: number | Date): number;
+}
+
+export class ClockBasedBlockTime implements BlockTime {
+  #getReferenceClockTime: () => number | Date;
+  #timeOffset: number = 0;
+
+  private static validateTimestamp(timestamp: number | Date) {
+    if (timestamp < 0) {
+      throw new Error(`Invalid timestamp: ${timestamp}. Value must be positive.`);
+    }
+  }
+
+  constructor(getReferenceClockTime: () => number | Date, startTime: number | Date | undefined) {
```

Comment:
I made the startTime explicitly accept `undefined`, and switched the argument order to make it more intuitive. I didn't make the argument optional, because the only time that we don't pass it (except for providing a variable of value `undefined`) is in tests.

This allows us to instantiate a `ClockBasedBlockTime` with no offset, without jumping through hoops.